pkgs/development/python-modules/sentencepiece/default.nix

```diff
@@ -0,0 +1,16 @@
+{ buildPythonPackage
+, stdenv
+, sentencepiece
+, pkgconfig
+}:
+
+buildPythonPackage rec {
+  pname = "sentencepiece";
+  inherit (sentencepiece) version src meta;
+
+  postPatch = ''
+    cd python
+  '';
+  buildInputs = [ sentencepiece ];
+  nativeBuildInputs = [ pkgconfig ];
```

Comment:
usually inputs come after src, and then ordered by more host-specific to target-specific inputs

here's how I would write this:
```
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ sentencepiece ];

  sourceRoot = "$sourceRoot/python"
```

Reply:
It doesn't seem to recognize "${sourceRoot}/python":
```
error: while evaluating the attribute 'buildInputs' of the derivation 'nix-shell' at /home/pash/nixpkgs/pkgs/build-support/mkshell/default.nix:28:3:
while evaluating the attribute 'passAsFile' of the derivation 'python3-3.7.5-env' at /home/pash/nixpkgs/pkgs/build-support/trivial-builders.nix:7:14:
while evaluating 'requiredPythonModules' at /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:64:27, called from /home/pash/nixpkgs/pkgs/development/interpreters/python/wrapper.nix:15:13:
while evaluating 'unique' at /home/pash/nixpkgs/lib/lists.nix:643:12, called from /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:66:6:
while evaluating 'requiredPythonModules' at /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:64:27, called from /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:82:33:
while evaluating 'unique' at /home/pash/nixpkgs/lib/lists.nix:643:12, called from /home/pash/nixpkgs/pkgs/top-level/python-packages.nix:66:6:
while evaluating 'unique' at /home/pash/nixpkgs/lib/lists.nix:643:12, called from /home/pash/nixpkgs/lib/lists.nix:649:17:
while evaluating anonymous function at /home/pash/nixpkgs/lib/lists.nix:152:16, called from undefined position:
while evaluating the attribute 'out.outPath' at /home/pash/nixpkgs/lib/customisation.nix:151:13:
while evaluating the attribute 'sourceRoot' of the derivation 'python3.7-sentencepiece-0.1.84' at /home/pash/nixpkgs/pkgs/development/interpreters/python/mk-python-derivation.nix:102:3:
infinite recursion encountered, at /home/pash/nixpkgs/pkgs/development/python-modules/sentencepiece/default.nix:14:19
```

looking into it

Reply:
you don't want to do `${sourceRoot}/python"; you just want `$sourceRoot/python`;

Reply:
`${sourceRoot}` needs to be evaluated by nix, but we are defining sourceRoot, so it's referencing itself, thus the infinite recursion. If we just do the bash interpolation, then it just gets evaluated once

Reply:
@jonringer yeah, i tried that too and it can't find the directory in that case: 

```
these derivations will be built:
  /nix/store/4k9rln87gj7sz91vh1hlgsqk4ard45gf-python3.7-sentencepiece-0.1.84.drv
  /nix/store/7q70rczzy68mfld2xh8cw4r80qci9s3p-python3.7-transformers-2.2.1.drv
  /nix/store/w3j539ak3h54sbfvp7w2r4gplag160h7-python3-3.7.5-env.drv
building '/nix/store/4k9rln87gj7sz91vh1hlgsqk4ard45gf-python3.7-sentencepiece-0.1.84.drv'...
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing setuptools-check-hook
Using setuptoolsCheckPhase
unpacking sources
unpacking source archive /nix/store/y7a0kjka6gdlfdrqq9i32g78lvsc2287-source
source root is $sourceRoot/python
chmod: cannot access '$sourceRoot/python': No such file or directory
builder for '/nix/store/4k9rln87gj7sz91vh1hlgsqk4ard45gf-python3.7-sentencepiece-0.1.84.drv' failed with exit code 1
cannot build derivation '/nix/store/w3j539ak3h54sbfvp7w2r4gplag160h7-python3-3.7.5-env.drv': 1 dependencies couldn't be built
error: build of '/nix/store/w3j539ak3h54sbfvp7w2r4gplag160h7-python3-3.7.5-env.drv' failed
```

Reply:
this works though: 

```
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ sentencepiece ];

  postUnpack = "sourceRoot=$sourceRoot/python";

```

Reply:
this also works: 

```
  nativeBuildInputs = [ pkgconfig ];
  buildInputs = [ sentencepiece ];

  sourceRoot = "source/python";

```

Reply:
I went with the latter... rebased and fixed.

Reply:
yea, I usually use the latter as well. I thought you would be able to use what I recommended, guess I should have tested first.

Reply:
actually, that makes sense, when it evaluates the derivation, that's when it will set the sourceRoot value, which doesn't exist, so if you do the bash substitution like i recommend it would just be empty so `sourceRoot="$sourceRoot/python"` would be just "/python", which isn't right. Sorry for leading you astray :(